fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/binders/LogicalMetricGenerator.java

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 Oath Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.web.apirequest.binders;
+
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.METRICS_NOT_IN_TABLE;
+
+import com.yahoo.bard.webservice.data.metric.LogicalMetric;
+import com.yahoo.bard.webservice.data.metric.MetricDictionary;
+import com.yahoo.bard.webservice.table.LogicalTable;
+import com.yahoo.bard.webservice.web.BadApiRequestException;
+import com.yahoo.bard.webservice.web.ErrorMessageFormat;
+
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Generates an ordered set of logical metrics based on a String representing the metrics requested by the user.
+ */
+public interface LogicalMetricGenerator {
+
+    /**
+     * Given a single dimension filter string, generate a metric name extension.
+     *
+     * @param filterString  Single dimension filter string.
+     *
+     * @return Metric name extension created for the filter.
+     */
+    static String generateMetricName(String filterString) {
+        return filterString.replace("|", "_").replace("-", "_").replace(",", "_").replace("]", "").replace("[", "_");
+    }
+
+    /**
+     * Extracts the list of metrics from the url metric query string and generates a set of LogicalMetrics.
+     * <p>
+     * If the query contains undefined metrics, {@link com.yahoo.bard.webservice.web.BadApiRequestException} will be
+     * thrown.
+     *
+     * @param apiMetricQuery  URL query string containing the metrics separated by ','
+     * @param metricDictionary  Metric dictionary contains the map of valid metric names and logical metric objects
+     *
+     * @return set of metric objects
+     */
+    LinkedHashSet<LogicalMetric> generateLogicalMetrics(String apiMetricQuery, MetricDictionary metricDictionary);
+
+    /**
+     * Validate that all metrics are part of the logical table.
+     *
+     * @param logicalMetrics  The set of metrics being validated
+     * @param table  The logical table for the request
+     *
+     * @throws BadApiRequestException if the requested metrics are not in the logical table
+     */
+    void validateMetrics(Set<LogicalMetric> logicalMetrics, LogicalTable table) throws BadApiRequestException;
+
+    /**
+     * Default implementation of this interface.
+     */
+    LogicalMetricGenerator DEFAULT_LOGICAL_METRIC_GENERATOR =
```

Comment:
Why make this an anonymous instance? 
I don't disagree, necessarily, but I want to understand the thinking.

Reply:
I think it's nice for the interface to provide the default implementations. I think it makes it easier to understand how the generators are used when the default implementation is directly linked through the interface, as it requires no knowledge of our coding practices (such as naming/packaging conventions) to find a good example of how the interfaces are intended to be used.

As for using an anonymous class, most of the default implementations are:
1. stateless, so the a single instance can be created here and used everywhere.
2. very small. LogicalMetricGenerator is an exception, but most of the default implementations of the generate methods are 5 lines or less. Making the default implementation an anonymous class greatly reduces clutter in most situations. While that is not true for LogicalMetricGenerator, I think the other benefits of having the anonymous class on the interface outweigh the ugliness of the long implementation being on the interface.

Additionally, using an anonymous class cuts down on the amount of classes we are making, which IMO makes it easier to navigate, and I think it makes it easier to understand how the interface is used. I think it's really easy and convenient for deciphering how the interface is supposed to be used when you can see the contract and the default implementation in one place.

Reply:
In this case it's definitely NOT that simple.  I think we should make the generate and validate into explicit static methods, which can be tested, and then make the anonymous class delegate to them.